### PR TITLE
adding shogun

### PR DESCRIPTION
**Pull request by antgonza**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -30,14 +30,18 @@ install:
   - pip install -U pip
   - conda install --yes -c jjhelmus glpk
   - conda install --yes pigz
-  - conda install --yes -c bioconda diamond=0.8 bowtie2 fastqc=0.11.5 samtools bedtools
+  - conda install --yes -c bioconda diamond=0.8 bowtie2 fastqc=0.11.5 samtools bedtools seqtk
+  - conda install --yes -c knights-lab shogun
   - pip install sphinx sphinx-bootstrap-theme coveralls biopython
   - pip install https://github.com/qiita-spots/qiita_client/archive/master.zip
   - pip install atropos
+  - pip install git+https://github.com/knights-lab/SHOGUN.git
```
Comment (antgonza) [inline .travis.yml]: Is this line really necessary? I thought the conda install `conda install --yes -c knights-lab shogun` will take care of this.

File: qp_shotgun/shogun/shogun.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    for seq_header, seq, qual_header, qual in itertools.zip_longest(*[fh] * 4):
+        yield (seq_header.strip(), seq.strip(), qual_header.strip(),
+               qual.strip())
+
+def generate_fna_file(fwd_seqs, rev_seqs, temp_path, map_file):
+    # Combines reverse and forward seqs per sample
+    # Returns filepaths of new combined files
+    both_fp = []
+    samples = make_read_pairs_per_sample(fwd_seqs, rev_seqs, map_file)
+    for run_prefix, sample, f_fp, r_fp in samples:
```
Comment (antgonza) [inline qp_shotgun/shogun/shogun.py]: You could replace for this and will generate a single FNA
```
    output_fp = join(temp_path, 'merged.fna')
    output = open(output_fp, "w")
    for run_prefix, sample, f_fp, r_fp in samples:
        counter = 0
        # Loop through forward file
        for seq_header, seq, qual_header, qual in _read_fastq_seqs(f_fp):
            output.write("%s_%d\n" % (sample, counter))
            output.write("%s\n" % seq)
            count+=1
        # Loop through reverse file
        for seq_header, seq, qual_header, qual in _read_fastq_seqs(r_fp):
            output.write("%s_%d\n" % (sample, counter))
            output.write("%s\n" % seq)
            count+=1
   output.close()
   return output_fp
```

File: .travis.yml (hunk 1 above)
Comment (semarpetrus) [reply .travis.yml]: Yeah that's shouldn't have been there. It was form an old draft.

File: qp_shotgun/shogun/shogun.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+    # Get the artifact metadata
+    prep_info = qclient.get('/qiita_db/prep_template/%s/'
+                            % artifact_info['prep_information'][0])
+    qiime_map = prep_info['qiime-map']
+
+    # Step 2 converting to fna
+    qclient.update_job_step(
+        job_id, "Step 2 of 6: Converting to FNA for Shogun")
+    temp_path = os.environ['QC_SHOGUN_TEMP_DP']
```
Comment (antgonza) [inline qp_shotgun/shogun/shogun.py]: Do we really need this new folder? Or can we simply use do something like:
`temp_path = join(out_dir, 'tmp')`
Note that the difference is that if we use QC_SHOGUN_TEMP_DP the folder is gonna be always created in that path, which is in home/where-the-software-was-installed but if we do it like in my suggestion the path is gonna be created in the mount/job-folder.

File: qp_shotgun/shogun/tests/test_shogun.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+            'shogun_burst': {
+                'Database': join(self.db_path, 'shogun'),
+                'Aligner tool': 'burst',
+                'Taxonomy Level': 'all',
+                'Number of threads': 1}}
+
+        self.assertEqual(obs, exp)
+
+    def test_generate_fna_file(self):
+        temp_path = os.environ['QC_SHOGUN_TEMP_DP']
```
Comment (antgonza) [inline qp_shotgun/shogun/tests/test_shogun.py]: For tests you could use: `from tempfile import mkdtemp` to avoid having to set this variable.

**Review by antgonza — COMMENTED** [review]
A multiple comment question.

File: qp_shotgun/shogun/shogun.py (hunk 3 above)
Comment (semarpetrus) [reply qp_shotgun/shogun/shogun.py]: The idea was that we can change the base temp path by changing the environment variable without having to reload the whole plugin. However, it makes sense to simply make the temp directory in the output

File: qp_shotgun/shogun/shogun.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        samples = make_read_pairs_per_sample(
+            fps['raw_forward_seqs'], rs, qiime_map)
+
+        comb_fp = generate_fna_file(temp_dir, samples)
+        # Combining files
+        parameters = _format_params(parameters, SHOGUN_PARAMS)
+        # Step 3 align
+        qclient.update_job_step(
+            job_id, "Step 3 of 6: Aligning FNA with Shogun")
+        generate_shogun_align_commands(
```
Comment (tanaes) [inline qp_shotgun/shogun/shogun.py]: This should return a `cmd`, right?

File: qp_shotgun/shogun/shogun.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        # Step 4 taxonomic profile
+        qclient.update_job_step(
+            job_id, "Step 4 of 6: Taxonomic profile with Shogun")
+        cmd, profile_fp = generate_shogun_assign_taxonomy_commands(
+            temp_dir, parameters)
+        # Step 5 functional profile
+        qclient.update_job_step(
+            job_id, "Step 5 of 6: Functional profile with Shogun")
+        levels = ['genus', 'species', 'level']
+        for level in levels:
```
Comment (tanaes) [inline qp_shotgun/shogun/shogun.py]: You'll also want to create redistributed Biom files for the taxonomy tables at these levels, a la: 

https://github.com/biocore/oecophylla/blob/251fa6c5a34014b919736acf3e003da13f09053b/oecophylla/taxonomy/taxonomy.rule#L378

File: qp_shotgun/shogun/tests/test_shogun.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+        self.assertEqual(obs_cmd, exp_cmd)
+
+    def test_generate_shogun_assign_taxonomy_commands(self):
+        out_dir = self.out_dir
+        with TemporaryDirectory(dir=out_dir, prefix='shogun_') as temp_dir:
+
+            exp_cmd = [
+                ('shogun assign_taxonomy --aligner bowtie2 '
+                 '--database %sshogun --input %s/alignment.bowtie2.sam '
```
Comment (tanaes) [inline qp_shotgun/shogun/tests/test_shogun.py:135]: I don't understand why `%sshogun`?

**Review by tanaes — COMMENTED** [review]
Looks pretty good, a few questions and suggestions.

I think in the current iteration (prior to separating out the taxonomy per-level biom generation as a separate step) you'll want to incorporate the 'regroup' command for the taxonomy tables as with the functional tables.

File: qp_shotgun/shogun/shogun.py (hunk 6 above)
Comment (tanaes) [reply qp_shotgun/shogun/shogun.py]: @semarpetrus The linked lines are the command you'll want to add to get the redistributed taxonomy tables. These are then what you'll convert to BIOM representation for the taxonomy portion of the command.

**Comment by antgonza** [comment]
The error is a timeout from the humman2 tests that we will move to a new repo. Thus, @semarpetrus, @tanaes and I decided to be ok to merge.

**Pull request merged**